File: ports/typescript-edge/src/telemetry.ts
/**
 * Bounded-fleet telemetry counters, transcribed from
 * integrations/scale-5m/aurora_scale/fleet.py (Telemetry).
 *
 * Every counter is observable; nothing about queue rejections or in-flight peaks is folded
 * into a generic "errors" bucket that would make a refusal look like work. The snapshot is
 * serialized through the canonical JSON serializer, so its key order is alphabetical and its
 * bytes are identical to the Python layer's sort_keys dumps - the property parity vectors pin.
 */

import { canonicalJsonString, type CanonicalValue } from "./canonical.js";

export interface TelemetrySnapshot {
  readonly completed: number;
  readonly dispatched: number;
  readonly in_flight: number;
  readonly lease_expiries: number;
  readonly peak_in_flight: number;
  readonly rejected_backpressure: number;
  readonly submitted: number;
}

export class Telemetry {
  private submittedCount = 0;
  private dispatchedCount = 0;
  private completedCount = 0;
  private rejectedBackpressureCount = 0;
  private leaseExpiriesCount = 0;
  private inFlight = 0;
  private peakInFlight = 0;

  /** Records one admitted dispatch. */
  dispatch(): void {
    this.submittedCount += 1;
    this.dispatchedCount += 1;
    this.inFlight += 1;
    if (this.inFlight > this.peakInFlight) this.peakInFlight = this.inFlight;
  }

  /**
   * Records one settled attempt. The clamp mirrors the reference: an over-counted completion
   * cannot drive in-flight negative and disguise itself as consistency.
   */
  complete(): void {
    this.completedCount += 1;
    if (this.inFlight > 0) this.inFlight -= 1;
  }

  /** Records one admission refused at a full queue. */
  rejectBackpressure(): void {
    this.rejectedBackpressureCount += 1;
  }

  /** Records one lease that lapsed to TTL. */
  leaseExpired(): void {
    this.leaseExpiriesCount += 1;
  }

  snapshot(): TelemetrySnapshot {
    return {
      completed: this.completedCount,
      dispatched: this.dispatchedCount,
      in_flight: this.inFlight,
      lease_expiries: this.leaseExpiriesCount,
      peak_in_flight: this.peakInFlight,
      rejected_backpressure: this.rejectedBackpressureCount,
      submitted: this.submittedCount,
    };
  }

  /**
   * Canonical bytes of the snapshot. The serializer sorts keys by code point, so byte-stability
   * does not depend on object literal order or engine map quirks.
   */
  snapshotJSON(): string {
    return canonicalJsonString(this.snapshot() as unknown as CanonicalValue);
  }
}

File: ports/go-control-plane/telemetry.go
package controlplane

import (
	"encoding/json"
)

// Telemetry accumulates bounded-fleet counters. Every counter is observable;
// nothing about queue rejections, lease expiries or in-flight peaks is folded
// into a generic "errors" bucket that would make a refusal look like work.
type Telemetry struct {
	submitted            uint64
	dispatched           uint64
	completed            uint64
	rejectedBackpressure uint64
	leaseExpiries        uint64
	inFlight             uint64
	peakInFlight         uint64
}

// TelemetrySnapshot is the exported shape of the counters. Fields are declared
// in alphabetical order on purpose: encoding/json emits struct fields in
// declaration order, so the JSON snapshot is byte-stable and its key order
// matches the Python scale layer's sort_keys dumps exactly.
type TelemetrySnapshot struct {
	Completed            uint64 `json:"completed"`
	Dispatched           uint64 `json:"dispatched"`
	InFlight             uint64 `json:"in_flight"`
	LeaseExpiries        uint64 `json:"lease_expiries"`
	PeakInFlight         uint64 `json:"peak_in_flight"`
	RejectedBackpressure uint64 `json:"rejected_backpressure"`
	Submitted            uint64 `json:"submitted"`
}

// Dispatch records one admitted dispatch.
func (t *Telemetry) Dispatch() {
	t.submitted++
	t.dispatched++
	t.inFlight++
	if t.inFlight > t.peakInFlight {
		t.peakInFlight = t.inFlight
	}
}

// Complete records one settled attempt. The clamp mirrors the reference: an
// over-counted completion cannot drive in-flight negative and disguise itself
// as consistency.
func (t *Telemetry) Complete() {
	t.completed++
	if t.inFlight > 0 {
		t.inFlight--
	}
}

// RejectBackpressure records one admission refused at a full queue.
func (t *Telemetry) RejectBackpressure() { t.rejectedBackpressure++ }

// LeaseExpired records one lease that lapsed to TTL.
func (t *Telemetry) LeaseExpired() { t.leaseExpiries++ }

// Snapshot returns the counters as a value.
func (t *Telemetry) Snapshot() TelemetrySnapshot {
	return TelemetrySnapshot{
		Completed:            t.completed,
		Dispatched:           t.dispatched,
		InFlight:             t.inFlight,
		LeaseExpiries:        t.leaseExpiries,
		PeakInFlight:         t.peakInFlight,
		RejectedBackpressure: t.rejectedBackpressure,
		Submitted:            t.submitted,
	}
}

// SnapshotJSON renders the snapshot with stable key order. Maps are avoided
// deliberately: Go randomizes map iteration, which would make byte-for-byte
// telemetry comparisons impossible.
func (t *Telemetry) SnapshotJSON() ([]byte, error) {
	return json.Marshal(t.Snapshot())
}
